liblangutil/SourceLocation.h

```diff
@@ -67,37 +68,46 @@ struct SourceLocation
 
 	bool isEmpty() const { return start == -1 && end == -1; }
 
+	bool hasSource() const noexcept { return charStream; }
+	bool hasSourceName() const noexcept { return charStream && !charStream->filename().empty(); }
+
+	std::string const& sourceName() const
+	{
+		static std::string const empty;
+		return charStream ? charStream->filename() : empty;
+	}
+
 	int start;
 	int end;
-	std::shared_ptr<std::string const> sourceName;
+	CharStream const* charStream; // non-owning, readonly
```

Comment:
I'm not sure if this is a good idea. With this change, all instances of SourceReference rely on the underlying CharStream's lifetime.

Reply:
Discussion resolution: use `shared_ptr<CharStream>`